src/compute/src/server.rs

```diff
@@ -217,6 +223,41 @@ impl mz_cluster::types::AsRunnableWorker<ComputeCommand, ComputeResponse> for Co
     }
 }
 
+/// Set the current thread's core affinity, based on the given `worker_id`.
+fn set_core_affinity(worker_id: usize) {
+    let Some(mut core_ids) = core_affinity::get_core_ids() else {
+        warn!(worker_id, "unable to get core IDs for setting affinity");
+        return;
+    };
+
+    // The `get_core_ids` docs don't say anything about a guaranteed order of the returned Vec,
+    // so sort it just to be safe.
+    core_ids.sort_unstable_by_key(|i| i.id);
+
+    // On multi-process replicas `worker_id` might be greater than the number of available cores.
+    // However, we assume that we always have at least as many cores as there are local workers.
+    // Violating this assumption is safe but might lead to degraded performance due to skew in core
+    // utilization.
+    let idx = worker_id % core_ids.len();
+    let core_id = core_ids[idx];
+
+    if core_affinity::set_for_current(core_id) {
+        info!(
+            worker_id,
+            core_id = core_id.id,
+            "set core affinity for worker"
+        );
+    } else {
+        // Setting core affinity is known to not work on Apple Silicon:
+        // https://github.com/Elzair/core_affinity_rs/issues/22.
+        warn!(
+            worker_id,
+            core_id = core_id.id,
+            "failed to set core affinity for worker"
```

Comment:
Can we upgrade `warn!` to `error!` in the case where we are _not_ on Apple Silicon? (If it's easy to detect that, which I'm not sure)

Reply:
Maybe we can just not bother trying to set the core affinity at all if we are on macOS. Does this also fail on Linux on Apple Sililcon, though?

Reply:
Yeah, I think adding an OS guard at the beginning of the function makes sense.

> Does this also fail on Linux on Apple Sililcon, though?

If you mean Linux in a VM, it works fine. If you mean something like Asahi Linux, I have no idea, but also I don't think we need to worry about that :)

Reply:
Yes, I meant in a VM (e.g. Docker). We will not have to worry about running Mz on Asahi for a while 🙂